### DISCUSSION TITLE
Display the received frame CRC?

### DISCUSSION THREAD

**Discussion by lansalot** [collin80/SavvyCAN#699] [Q&A]
Afternoon all - did log this as an issue, but nothing doing there; discussions seems to be the place to be!

Been looking but can't see any way to display the CRC as received for the CAN frame?

This would be useful as I've a system I'm trying to work out how the CRC is calculated - but without seeing the CRCs from legitimate packets, it's proving difficult and we're having to use non-SavvyCAN stuff to capture the frames and print them out. SavvyCAN being the tool of choice, it would be great to have them in the output?

Hopefully I'm just missing something obvious, but it would make a great enhancement if not?

**Comment by cschweers** [comment]
As this is done by the CAN bus hardware itself and not by SavvyCAN, there is generally no CRC information available at application level.

However, there is code that shows how to calculate the CRC, for example:
https://github.com/KevinOConnor/can2040/blob/master/src/can2040.c

**Reply by collin80** [reply]
Basically this. I may be able to grab the CRC from socketcan, I'm not sure. I know I don't send it for GVRET devices. Most people don't care about the CAN level CRC as it is irrelevant most of the time. If it matches, the frame comes in. If it doesn't match, it is rejected by the hardware. It is possible for the hardware to notify higher protocol levels about this but usually the CAN bitstream CRC is more or less ignored by everything but the hardware on each node.

**Comment by lansalot** [comment]
Ah OK, thanks for that - I've received some trace files from a particular device where the CRC was included, but the reason I'd asked was because it appeared they're not using a standard CRC. Am trying to reverse-engineer it but of course it's proving difficult without that actual CRC in a trace of my own.

That conversation is over here: https://reverseengineering.stackexchange.com/questions/32093/help-reverse-engineering-a-crc

**Reply by esaulenka** [reply]
Sorry for a late answer, but you are talking about different things.

All CAN frames should contain CRC field. This CRC always calculated in the same way, described in the CAN specification (you can read it, ``Bosch CAN specification`, it is free). Typically this particular field generated and processed on the hardware level (and I am not sure that any MCUs allows to modify it somehow). At most, you can see it is correct or not. You cannot use own CRC algo there, it will break all communication with other devices.

And now you are asking about some custom checksum (most probably, CRC) implemented on top of CAN layer. There is no any common standard, all manufacturers using own (secret) algorithms. You can ask about it on StackExchange, [Reddit](https://www.reddit.com/r/CarHacking/) or search [CommaAI DBC base](https://github.com/commaai/opendbc)
